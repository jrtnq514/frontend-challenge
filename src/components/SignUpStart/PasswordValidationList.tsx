import React from "react";
import {
  Card,
  CardHeader,
  CardBody,
  Text,
  List,
  ListItem,
  ListIcon,
} from "@chakra-ui/react";
import { CheckIcon, CloseIcon } from "@chakra-ui/icons";
import { FieldErrorsImpl } from "react-hook-form";
import { SignUpStartInputs } from "src/views/SignUpStart";

interface PasswordValidationListItemProps {
  isError: boolean;
  children: React.ReactNode;
}

const PasswordValidationListItem: React.FC<PasswordValidationListItemProps> = ({
  isError,
  children,
}) => {
  return (
    <ListItem color={isError ? "red.500" : "green.700"} display="flex">
      {isError ? <ListIcon as={CloseIcon} /> : <ListIcon as={CheckIcon} />}
      {children}
    </ListItem>
  );
};

interface PasswordValidationListProps {
  errors: FieldErrorsImpl<SignUpStartInputs>;
}

const PasswordValidationList: React.FC<PasswordValidationListProps> = ({
  errors,
}) => {
  return (
    <Card width="100%">
      <CardHeader>
        <Text fontWeight={600} as="label">Passwords must:</Text>
      </CardHeader>
      <CardBody pt={0}>
        <List spacing={0.5}>
          <PasswordValidationListItem
            isError={
              !!(
                errors.password?.types?.minLength ||
                errors.password?.types?.required
              )
            }
          >
            Be at least 8 characters long
          </PasswordValidationListItem>
          <PasswordValidationListItem
            isError={!!errors.password?.types?.passwordUppercase}
          >
            Contain a uppercase letter
          </PasswordValidationListItem>
          <PasswordValidationListItem
            isError={!!errors.password?.types?.passwordLowercase}
          >
            Contain a lowercase letter
          </PasswordValidationListItem>
          <PasswordValidationListItem
            isError={!!errors.password?.types?.passwordNumber}
          >
            Contain a number
          </PasswordValidationListItem>
          <PasswordValidationListItem
            isError={!!errors.password?.types?.passwordSpecial}
          >
            Contain a special character
          </PasswordValidationListItem>
        </List>
      </CardBody>
    </Card>
  );
};

export default PasswordValidationList;
