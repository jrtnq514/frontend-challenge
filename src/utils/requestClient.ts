import ky from "ky";

const API_BASE_URL = process.env.API_BASE_URL || "http://localhost:3001/api";

export default ky.create({
  headers: {},
  prefixUrl: API_BASE_URL,
});
