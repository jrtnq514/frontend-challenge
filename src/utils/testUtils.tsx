import { InitialEntry, RouterState } from "@remix-run/router";
import {
  createMemoryRouter,
  RouteObject,
  RouterProvider,
} from "react-router-dom";
import {
  SignUpState,
  SignUpStateProvider,
} from "src/contexts/signUpStateContext";
import { render } from "@testing-library/react";

interface RenderViewProps {
  routes: RouteObject[];
  routerOptions?:
    | {
        basename?: string | undefined;
        hydrationData?:
          | Partial<Pick<RouterState, "errors" | "loaderData" | "actionData">>
          | undefined;
        initialEntries?: InitialEntry[] | undefined;
        initialIndex?: number | undefined;
      }
    | undefined;
  signUpState?: SignUpState;
}

export const renderView = ({
  routes,
  routerOptions = {},
  signUpState,
}: RenderViewProps) => {
  const router = createMemoryRouter(routes, routerOptions);

  return {
    ...render(
      <SignUpStateProvider initState={signUpState}>
        <RouterProvider router={router} />
      </SignUpStateProvider>
    ),
    router,
  };
};
