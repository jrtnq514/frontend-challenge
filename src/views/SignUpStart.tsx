import {
  Box,
  Heading,
  VStack,
  Text,
  FormLabel,
  FormControl,
  Input,
  HStack,
  Button,
  FormErrorMessage,
  InputRightElement,
  InputGroup,
  Popover,
  PopoverArrow,
  PopoverContent,
  PopoverCloseButton,
  PopoverHeader,
  PopoverBody,
  IconButton,
  useDisclosure,
  PopoverAnchor,
  List,
  ListItem,
  Spinner,
} from "@chakra-ui/react";
import { InfoIcon } from "@chakra-ui/icons";
import React, { useEffect, useState } from "react";
import { useNavigate, useNavigation } from "react-router-dom";
import { useSignUpState } from "src/contexts/signUpStateContext";
import { useForm, SubmitHandler } from "react-hook-form";
import PasswordValidationList from "src/components/SignUpStart/PasswordValidationList";

export interface SignUpStartInputs {
  firstName: string;
  email: string;
  password: string;
}

const SignUpStart: React.FC = () => {
  const navigate = useNavigate();
  const { state: navState } = useNavigation();
  const [showPassword, setShowPassword] = useState(false);
  const { isOpen, onToggle, onClose } = useDisclosure();
  const { signUpState, setSignUpState } = useSignUpState();
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm<SignUpStartInputs>({ criteriaMode: "all" });

  const passwordUppercase = (input: string) => /[A-Z]/.test(input);
  const passwordLowercase = (input: string) => /[a-z]/.test(input);
  const passwordNumber = (input: string) => /[0-9]/.test(input);
  const passwordSpecial = (input: string) =>
    /[~`! @#$%^&*()_\-+={[}\]|\:;"'<,>.?/]/.test(input);

  useEffect(() => {
    if (signUpState.firstName) {
      setValue("firstName", signUpState.firstName);
    }

    if (signUpState.email) {
      setValue("email", signUpState.email);
    }

    setSignUpState({
      ...signUpState,
      firstName: "",
      email: "",
      password: "",
    });
  }, []);

  useEffect(() => {
    if (
      signUpState.firstName &&
      signUpState.email &&
      signUpState.password
    ) {
      navigate("/more-info");
    }
  }, [signUpState]);

  const onSubmit: SubmitHandler<SignUpStartInputs> = (data) => {
    setSignUpState({
      ...signUpState,
      firstName: data.firstName,
      email: data.email,
      password: data.password,
    });
  };

  return (
    <Box>
      {navState === "loading" ? (
        <Spinner size="xl" />
      ) : (
        <VStack spacing={6}>
          <Heading as="h1">Sign Up</Heading>
          <Text>Lets get started with some basic information.</Text>
          <Box
            as="form"
            width="100%"
            data-id="start-form"
            onSubmit={handleSubmit(onSubmit)}
          >
            <VStack spacing={4}>
              <FormControl isInvalid={!!errors.firstName}>
                <FormLabel>First Name</FormLabel>
                <Input
                  data-id="first-name-input"
                  type="text"
                  {...register("firstName", { required: true })}
                />
                {errors.firstName && errors.firstName.types?.required && (
                  <FormErrorMessage>First Name is required.</FormErrorMessage>
                )}
              </FormControl>
              <FormControl isInvalid={!!errors.email}>
                <FormLabel>Email</FormLabel>
                <Input
                  data-id="email-input"
                  type="email"
                  {...register("email", { required: true })}
                />
                {errors.email && errors.email.types?.required && (
                  <FormErrorMessage>Email is required.</FormErrorMessage>
                )}
              </FormControl>
              <FormControl isInvalid={!!errors.password}>
                <FormLabel>
                  Password{" "}
                  <Popover isOpen={isOpen} onClose={onClose} placement="auto">
                    <PopoverAnchor>
                      <IconButton
                        onClick={onToggle}
                        aria-label="Show password requirements"
                        height="auto"
                        minW="initial"
                        icon={<InfoIcon />}
                      />
                    </PopoverAnchor>
                    <PopoverContent>
                      <PopoverArrow />
                      <PopoverCloseButton />
                      <PopoverHeader>Passwords must:</PopoverHeader>
                      <PopoverBody>
                        <List>
                          <ListItem>Be at least 8 characters long</ListItem>
                          <ListItem>Contain a uppercase letter</ListItem>
                          <ListItem>Contain a lowercase letter</ListItem>
                          <ListItem>Contain a number</ListItem>
                          <ListItem>Contain a special character</ListItem>
                        </List>
                      </PopoverBody>
                    </PopoverContent>
                  </Popover>
                </FormLabel>
                <InputGroup>
                  <Input
                    data-id="password-input"
                    type={showPassword ? "text" : "password"}
                    pr={12}
                    {...register("password", {
                      required: true,
                      minLength: 8,
                      validate: {
                        passwordUppercase,
                        passwordLowercase,
                        passwordNumber,
                        passwordSpecial,
                      },
                    })}
                  />
                  <InputRightElement width={12}>
                    <Button
                      size="sm"
                      variant="ghost"
                      onClick={(event) => setShowPassword(!showPassword)}
                    >
                      {showPassword ? "Hide" : "Show"}
                    </Button>
                  </InputRightElement>
                </InputGroup>
                <FormErrorMessage>
                  <PasswordValidationList errors={errors} />
                </FormErrorMessage>
              </FormControl>
              <HStack width="100%" justifyContent="end">
                <Button type="submit" variant="solid" colorScheme="green">
                  Next
                </Button>
              </HStack>
            </VStack>
          </Box>
        </VStack>
      )}
    </Box>
  );
};

export default SignUpStart;
