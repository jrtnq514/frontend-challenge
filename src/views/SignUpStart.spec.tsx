import { renderView } from "src/utils/testUtils";
import SignUpStart from "./SignUpStart";
import { waitFor, screen, fireEvent } from "@testing-library/react";

describe("SignUpStart", () => {
  it("should render three inputs", async () => {
    const { container } = renderView({
      routes: [
        {
          path: "/",
          element: <SignUpStart />,
        },
      ],
      routerOptions: { initialEntries: ["/"] },
    });

    await waitFor(() =>
      Promise.all([
        screen.findByTestId("first-name-input"),
        screen.findByTestId("email-input"),
        screen.findByTestId("password-input"),
      ])
    );

    expect(await screen.findByTestId("first-name-input")).not.toBe(null);
    expect(await screen.findByTestId("email-input")).not.toBe(null);
    expect(await screen.findByTestId("password-input")).not.toBe(null);
    expect(container).toMatchSnapshot();
  });

  it("should pre-populate name and email, if set in state", async () => {
    renderView({
      routes: [
        {
          path: "/",
          element: <SignUpStart />,
        },
      ],
      routerOptions: { initialEntries: ["/"] },
      signUpState: { firstName: "test", email: "test@test.com" },
    });

    await waitFor(() =>
      Promise.all([
        screen.findByTestId("first-name-input"),
        screen.findByTestId("email-input"),
      ])
    );

    const nameInput = (await screen.findByTestId(
      "first-name-input"
    )) as HTMLInputElement;
    const emailInput = (await screen.findByTestId(
      "email-input"
    )) as HTMLInputElement;

    expect(nameInput.value).toBe("test");
    expect(emailInput.value).toBe("test@test.com");
  });

  it("should display errors for invalid inputs on next click", async () => {
    renderView({
      routes: [
        {
          path: "/",
          element: <SignUpStart />,
        },
      ],
      routerOptions: { initialEntries: ["/"] },
    });

    await waitFor(() => Promise.all([screen.findByTestId("next-btn")]));

    const nextBtn = await screen.findByTestId("next-btn");
    fireEvent.click(nextBtn);

    expect(await screen.findByTestId("first-name-input-error")).not.toBe(null);
    expect(await screen.findByTestId("email-input-error")).not.toBe(null);
    expect(await screen.findByTestId("password-input-error")).not.toBe(null);
  });
});
