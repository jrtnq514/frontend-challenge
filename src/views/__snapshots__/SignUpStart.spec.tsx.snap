// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpStart should render three inputs 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="chakra-stack css-pfcv3f"
    >
      <h1
        class="chakra-heading css-0"
      >
        Sign Up
      </h1>
      <p
        class="chakra-text css-0"
      >
        Lets get started with some basic information.
      </p>
      <form
        class="css-8atqhb"
        data-testid="start-form"
      >
        <div
          class="chakra-stack css-wulyol"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="field-:r1:"
              id="field-:r1:-label"
            >
              First Name
            </label>
            <input
              class="chakra-input css-0"
              data-testid="first-name-input"
              id="field-:r1:"
              name="firstName"
              type="text"
            />
          </div>
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="field-:r2:"
              id="field-:r2:-label"
            >
              Email
            </label>
            <input
              class="chakra-input css-0"
              data-testid="email-input"
              id="field-:r2:"
              name="email"
              type="email"
            />
          </div>
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="field-:r3:"
              id="field-:r3:-label"
            >
              Password
               
              <button
                aria-label="Show password requirements"
                class="chakra-button css-xvp17x"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon css-onkibi"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
                    fill="currentColor"
                  />
                </svg>
              </button>
              <div
                class="chakra-popover__popper css-0"
                style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
              >
                <section
                  aria-describedby="popover-body-:r5:"
                  aria-labelledby="popover-header-:r5:"
                  class="chakra-popover__content css-1ab1jlb"
                  id="popover-content-:r5:"
                  role="dialog"
                  style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
                  tabindex="-1"
                >
                  <div
                    class="chakra-popover__arrow-positioner css-0"
                    data-popper-arrow=""
                    style="position: absolute;"
                  >
                    <div
                      class="chakra-popover__arrow css-0"
                      data-popper-arrow-inner=""
                    />
                  </div>
                  <button
                    aria-label="Close"
                    class="chakra-popover__close-btn css-6sdaxy"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="chakra-icon css-onkibi"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                        fill="currentColor"
                      />
                    </svg>
                  </button>
                  <header
                    class="chakra-popover__header css-0"
                    id="popover-header-:r5:"
                  >
                    Passwords must:
                  </header>
                  <div
                    class="chakra-popover__body css-0"
                    id="popover-body-:r5:"
                  >
                    <ul
                      class="css-155za0w"
                      role="list"
                    >
                      <li
                        class="css-0"
                      >
                        Be at least 8 characters long
                      </li>
                      <li
                        class="css-0"
                      >
                        Contain a uppercase letter
                      </li>
                      <li
                        class="css-0"
                      >
                        Contain a lowercase letter
                      </li>
                      <li
                        class="css-0"
                      >
                        Contain a number
                      </li>
                      <li
                        class="css-0"
                      >
                        Contain a special character
                      </li>
                    </ul>
                  </div>
                </section>
              </div>
            </label>
            <div
              class="chakra-input__group css-4302v8"
            >
              <input
                class="chakra-input css-1bpvmlm"
                data-testid="password-input"
                id="field-:r3:"
                name="password"
                type="password"
              />
              <div
                class="chakra-input__right-element css-dku7lc"
              >
                <button
                  class="chakra-button css-4xx2wk"
                  type="button"
                >
                  Show
                </button>
              </div>
            </div>
          </div>
          <div
            class="chakra-stack css-1tphubk"
          >
            <button
              class="chakra-button css-4xx2wk"
              data-testid="next-btn"
              type="submit"
            >
              Next
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
